go/types/graph_builder.go

```diff
@@ -0,0 +1,263 @@
+	b.oc.GraphListAppend(keys, v)
+}
+
+type GraphOpContainer struct {
+	keys []Value
+	kind NomsKind
+	item sequenceItem
+}
+
+// Builds and returns the graph.
+func (b *GraphBuilder) Build() Value {
+	d.Chk.True(b.oc != nil, "Can only call Finish() once")
+	defer func() {
+		b.oc = nil
+	}()
+
+	iter := b.oc.NewIterator()
+	defer iter.Release()
+	keyCnt := int64(0)
+
+	// start up a go routine that will do the reading from graphBuilder's private
+	// ldb opCache.
+	graphOpChan := make(chan GraphOpContainer, 512)
+	go func() {
+		for iter.Next() {
+			keys, kind, item := iter.GraphOp()
+			container := GraphOpContainer{keys: keys, kind: kind, item: item}
+			graphOpChan <- container
+		}
+		close(graphOpChan)
+	}()
+
+	// iterator returns keys, in sort order by array
+	for goc := range graphOpChan {
+		keys, kind, item := goc.keys, goc.kind, goc.item
+
+		// Get index of first key that is different than what is on the stack
+		idx := findDiff(b.stack, keys)
+
+		if idx == -1 {
+			// no keys have changed we're working on same coll as previous
+			// iteration, just append to sequenceChunker at top of stack
+			b.appendItemToCurrentTopOfStack(kind, item)
+		} else {
+			// Some keys that were in the last graphOp are no longer present
+			// which indicates that we are finished adding to those cols. Pop
+			// those keys from the stack. This will cause any popped cols to be
+			// closed and added to their parents.
+			for idx < b.stack.len()-1 {
+				b.popKeyFromStack()
+			}
+
+			// We may have popped some keys off of the stack and are left with
+			// an item to append to the stack of a previously existing key.
+			if b.stack.len()-1 == len(keys) {
+				b.appendItemToCurrentTopOfStack(kind, item)
+			}
+
+			// Or we may have some new keys to add to the stack. Add those keys
+			// and then append the item to the top element.
+			for b.stack.len()-1 < len(keys) {
+				k := kind
+				i := item
+				if b.stack.len()-1 < len(keys)-1 {
+					k = MapKind
+					i = nil
+				}
+				b.pushNewKeyOnStack(keys[b.stack.len()-1], k, i)
+				if b.verbose {
+					keyCnt++
+					status.Printf("Added %s keys to graph", humanize.Comma(keyCnt))
+				}
+			}
+		}
+	}
+
+	// We're done adding elements. Pop any intermediate keys off the stack and
+	// fold their results into their parent map.
+	for b.stack.len() > 1 {
+		b.popKeyFromStack()
+	}
+	res := b.stack.pop().done()
+	if b.verbose {
+		status.Done()
+	}
+	return res
+
+}
+
+// pushNewKeyOnStack() creates a new graphStackElem node and pushes it on the
+// stack. The new element contains the |key| and a new sequenceChunker that will
+// be appended to to build this node in the graph.
+func (b *GraphBuilder) pushNewKeyOnStack(key Value, kind NomsKind, item sequenceItem) {
+	var seq *sequenceChunker
+	switch kind {
+	case MapKind:
+		seq = newEmptySequenceChunker(b.vrw, b.vrw, makeMapLeafChunkFn(b.vrw), newOrderedMetaSequenceChunkFn(MapKind, b.vrw), mapHashValueBytes)
```

Comment:
can we abstract these, e.g. `newListChunker`, `newSetChunker`, `newMapChunker` to avoid code duplication with other places that are constructing appropriate chunkers?